apache2/waf_logging/waf_log_util.cc

```diff
@@ -1,122 +1,253 @@
+        const char* details_file, const char* details_line, const char* hostname, const char* waf_unique_id) {
     
     try {
-        // Set Waf format.
-        json_string = get_json_log_message(resourceId, operationName, category, instanceId, clientIP, clientPort, requestUri, ruleSetType, ruleSetVersion, ruleId, messages, action, site, details_messages, details_data, details_file, details_line, hostname, waf_unique_id); 
-    }
-    catch (...) {
-        return WAF_LOG_UTIL_FAILED;
-    }
-
-    *result_json = strdup((json_string + "\n").c_str());
-    return WAF_LOG_UTIL_SUCCESS; 
+        const std::string json_string = get_json_log_message(resource_id, operation_name, category,
+                instance_id, client_ip, client_port, request_uri, rule_set_type, rule_set_version,
+                rule_id, messages, action, site, details_messages, details_data, details_file,
+                details_line, hostname, waf_unique_id);
+        char* result = static_cast<char*>(std::malloc(json_string.length() + 2)); // Two extra bytes for \n and \0
+        std::copy(json_string.begin(), json_string.end(), result);
+        result[json_string.length()] = '\n';
+        result[json_string.length() + 1] = '\0';
+        return result;
+    } catch (...) {}
+
+    return nullptr;
 }
 
 void free_json(char* str) {
-    free(str);
+    std::free(str);
 }
 
-void init_appgw_rules_id_hash() {
-	ifstream infile(RULES_ID_FILE);
-	string line;
-
-	while (getline(infile, line)) {
-		int rule_id = stoi(line);
-		appgw_ruleid_hash[rule_id] = true;
-	}
-	infile.close();
+static rule_ids_vector read_disableable_rule_ids(const std::string& filename) {
+    rule_ids_vector rule_ids;
+    std::ifstream ifs(filename);
+    if(ifs) {
+        std::copy(std::istream_iterator<int>(ifs),
+                std::istream_iterator<int>(),
+                std::back_inserter(rule_ids));
+        std::sort(std::begin(rule_ids), std::end(rule_ids));
+    }
 
-	return;
+    return rule_ids;
 }
 
-bool rule_is_mandatory(int rule_id) {
-	return (appgw_ruleid_hash.find(rule_id) == appgw_ruleid_hash.end());
-}
\ No newline at end of file
+void init_appgw_rules_id_hash() {
+    std::map<rule_set_info, std::string> pathsMap {
+        {OWASP_2_2_9, "/etc/nginx/modsec/OWASP/2.2.9/disableableRules.conf"},
+        {OWASP_3_0, "/etc/nginx/modsec/OWASP/3.0/disableableRules.conf"},
+        {OWASP_3_1, "/etc/nginx/modsec/OWASP/3.1/disableableRules.conf"},
+    };
+    for (const auto& item : pathsMap) {
+        auto vec = read_disableable_rule_ids(item.second);
+        if (!vec.empty()) {
+            disableable_rule_ids_map[item.first] = vec;
+        }
+    }
+    if (disableable_rule_ids_map.size() > 0) {
+        // If we have found type/version-specific files, we don't need to read the default (deprecated one)
+        return;
+    }
+    default_disableable_rule_ids = read_disableable_rule_ids("/etc/nginx/modsec/disableableRules.conf");
```

Comment:
Should we add error checks around this file being present and not empty?

Reply:
That's best achieved through the use of `std::optional`. Let me switch to using it because I feel this is the right idiom for expressing the difference between something non-existent and empty.
(Technically, we don't care that much whether the file does not exist or it is just empty but doesn't hurt to be specific).

Reply:
Sounds good